TCP2: Simplify TCP packet creation

rveerama1 opened the pull request:
There are existing calls to create net packet based on context, create IP header and finalize the packet before sending through net_send_data(). Test protocol also changed as well to verify with TTCN3.

rveerama1 committed:
net: context: Move asserts to proper place

net_context local sin/sin6 addresses are used only
when input src is not available. So ASSERT should
be inside if block.

Signed-off-by: Ravi kumar Veeramally <ravikumar.veeramally@linux.intel.com>

tbursztyka commented on subsys/net/ip/tcp2.c:
```diff
@@ -465,113 +466,122 @@ static size_t tcp_data_get(struct tcp *conn, struct net_pkt *pkt)
 }
 
-static void ip_header_add(struct tcp *conn, struct net_pkt *pkt)
+static int tcp_header_add(struct tcp *conn, struct net_pkt *pkt, u8_t flags)
```
call it tcp_create() to stay with the same naming we have all over the net stack.

tbursztyka commented on subsys/net/ip/tcp2.c:
```diff
@@ -465,113 +466,122 @@ static size_t tcp_data_get(struct tcp *conn, struct net_pkt *pkt)
-		ip6.vtc = 0x60;
-		ip6.nexthdr = IPPROTO_TCP;
-		ip6.len = htons(net_pkt_get_len(pkt));
+static int ip_header_add(struct tcp *conn, struct net_pkt *pkt)
```
hopefully this will go away once tcp2 will be properly integrated with net_context

rveerama1 replied:
yep.

tbursztyka commented on subsys/net/ip/tp.c:
```diff
@@ -8,6 +8,8 @@
 #include <stdlib.h>
 #include "tp.h"
 #include "tp_priv.h"
+#include "ipv4.h"
```
Could you split your patch in 2? tp is not tcp2.

rveerama1 replied:
done

tbursztyka commented on subsys/net/ip/tp.c:
```diff
@@ -448,14 +424,36 @@ static void tp_pkt_send(struct net_pkt *pkt)
 	struct net_buf *buf;
 
-	pkt->family = af;
+	ip_header_add(pkt);
```
why not using net_ipv4_create()/net_ipv6_create() ?

rveerama1 replied:
ip_header_add does that. it calls net_ipv4_create and net_ipv6_create().

ozhuraki reviewed (approved):
Thanks and LGTM

tbursztyka commented on subsys/net/ip/tp.c:
```diff
@@ -448,14 +424,36 @@ static void tp_pkt_send(struct net_pkt *pkt)
-	pkt->family = af;
+	ip_header_add(pkt);
+
+	udp_header_add(pkt);
```
same with net_udp_create()

rveerama1 replied:
done

tbursztyka commented on subsys/net/ip/tp.c:
```diff
@@ -448,14 +424,36 @@ static void tp_pkt_send(struct net_pkt *pkt)
 		size_t data_len, const char *file, int line)
 {
-	struct net_pkt *pkt = tp_pkt_alloc(0, file, line);
+	struct net_pkt *pkt = tp_output_pkt_alloc(af, iface, file, line);
```
you know the data length, go ahead: pass it to tp_output_pkt_alloc() and make it allocate for a udp hdr + the data length.

tbursztyka commented on subsys/net/ip/tp.c:
```diff
@@ -465,14 +463,10 @@ void _tp_output(sa_family_t af, struct net_if *iface, void *data,
 		memcpy(net_buf_add(buf, len), data, len);
 
-		net_pkt_frag_add(pkt, buf);
+		net_pkt_append_buffer(pkt, buf);
```
and then here get rid of the for () loop and us net_pkt_writg(pkt, data, len). Done.

rveerama1 replied:
done

jukkar commented on subsys/net/ip/tcp2.c:
```diff
@@ -465,113 +466,122 @@ static size_t tcp_data_get(struct tcp *conn, struct net_pkt *pkt)
+{
+	if (IS_ENABLED(CONFIG_NET_IPV4) && net_pkt_family(pkt) == AF_INET) {
+		return net_context_create_ipv4_new(conn->context, pkt,
+						&(conn->src->sin.sin_addr),
```
No need to put () after &, so `&conn->src->sin.sin_addr` is enough. Please fix all of them in this PR.

rveerama1 replied:
done

rveerama1 committed:
net: tcp2: Simplify TCP pkt creation

Use existing network stack calls to create net pkt,
ip header and tcp header creation. Also simplify
finalize api calls.

Signed-off-by: Ravi kumar Veeramally <ravikumar.veeramally@linux.intel.com>

rveerama1 committed:
net: tcp2: tp: Simplify pkt creation in test protocol

Use existing network stack calls to create net pkt,
ip header and udp header creation. Also simplify
finalize api calls.

Signed-off-by: Ravi kumar Veeramally <ravikumar.veeramally@linux.intel.com>
